crates/ruff_db/src/files.rs

```diff
@@ -131,6 +126,44 @@ impl Files {
         Some(file)
     }
 
+    /// Refreshes the state of all known files under `path` recursively.
+    ///
+    /// The most common use case is to update the [`Files`] state after removing or moving a directory.
+    ///
+    /// # Performance
+    /// Refreshing the state of every file under `path` is expensive. It requires iterating over all known files
+    /// and making system calls to get the latest status of each file in `path`.
+    /// That's why [`File::sync_path`] and [`File::sync_path`] is preferred if it is known that the path is a file.
+    #[tracing::instrument(level = "debug", skip(db))]
+    pub fn sync_recursively(db: &mut dyn Db, path: &SystemPath) {
```

Comment:
We may want to optimize this in the future, but I have decided against it for now. I suspect that the cost of iterating all files is neglectable compared to the necessary sys-calls to read the file metadata.

We'll see once Red Knot is capable of analyzing real world projects.